### PR TITLE
refactor(core): unify mesh picking methods

### PR DESCRIPTION
**Pull request by peppsac**
The idea of this PR is to expose a single entry point for querying the view.

This method belongs to the View.

Here are a few example usage:
  * view.pickObjectsAt({ x, y })
  * view.pickObjectsAt({ x, y }, 'wfsBuilding')
  * view.pickObjectsAt({ x, y }, 'wfsBuilding', myLayer)
  * view.pickObjectsAt({ x, y }, 'wfsBuilding', myObject3d)

It supports either layers, layer's id or THREE.Object3D.
By default it'll use THREE.Raycaster, but some layer can provide their own implementation.

Currently the layer with a custom implementation are:
  - layers using TileMesh (globe, planar, panorama)
  - pointclouds

### CODE REVIEW

File: src/Core/Scheduler/Providers/PointCloudProvider.js
Comment (gchoqueux) [inline src/Core/Scheduler/Providers/PointCloudProvider.js]: Good refacto and unification.
However I have an objection on the choice of files for the picking declaration : the providers 

To be more generic, clearer and to reuse,  should be in a picking toolbox.
These tools must have as input the type of object to pick.

Then the layer's picking functionality must be assigned to its construction, but not in the pre-process phase of the provider. Because the provider must take care to supply only what has in title and pre-process works only for this supply.

it's same case in PR #546 
what is your motivation to put this in the providers?

I also ask the opinion to  @mbredif ?

Comment (peppsac) [reply src/Core/Scheduler/Providers/PointCloudProvider.js]: > To be more generic, clearer and to reuse, should be in a picking toolbox.
> [...]
>  what is your motivation to put this in the providers?

Yes; we already had this discussion earlier, so I won't copy-paste all my arguments from there.

I can move `pickObjectsAt` to a `Picking.js` file if it you prefer.

But I still believe that `preprocessing` of geometry Providers is a good place to extend layer's object so I'll leave the `layer.pickObjectsAt` there.

Comment (peppsac) [reply src/Core/Scheduler/Providers/PointCloudProvider.js]: Done.

File: src/Core/Scheduler/Providers/TileProvider.js
Comment (gchoqueux) [inline src/Core/Scheduler/Providers/TileProvider.js]: All references to the layer should disappear from `tileProvider`. Because there isn't reason for a tile provider to know the layers. For example, in the future, remove any notion of layer in `TileProvider.prototype.executeCommand`. You have to be able to use the tile provider without a layer concept. 

`preprocessDataLayer` also has no reason to be here and even less if that add functions that aren't `tile building`

So don't add new reverence to the layer

**Review by gchoqueux — CHANGES_REQUESTED** [review]

File: src/Core/Scheduler/Providers/TileProvider.js
Comment (autra) [reply src/Core/Scheduler/Providers/TileProvider.js]: > Because there isn't reason for a tile provider to know the layers

Every single provider of itowns absolutely *does* know about layers. Providers are here to interface layers with 2 things: 
- a protocol by:
    - initializing the layer with some default values for this protocol and / or metadata from the network, following the protocol specifications (getCapabilities, tileset.json...)
    - abstract the fetching of data from servers implementing this protocol
- The scheduler mechanism, because we're fetching datas from the internet. And it uses the layer because it provides some configurations that `executeCommand` needs to generate new nodes (opacity, colors, overrideMaterials etc...)

And in turns, the way we create node for a given protocol have an influence on how we should pick it on the scene. So it makes a lot of sense that this logic should be at least partly in providers.

> preprocessDataLayer also has no reason to be here

For the aforementioned reason, it has all the reason to be in a provider

Not sure I understand you well. Could you describe what alternative solution you would choose? Where would you move all that code then? It might make things clearer.

Comment (peppsac) [reply src/Core/Scheduler/Providers/TileProvider.js]: Moved more code so now TileProvider isn't modified.

Is this what you had in mind @gchoqueux ?

Comment (gchoqueux) [reply src/Core/Scheduler/Providers/TileProvider.js]: The best place for picking is the construction of the geometric layer using the three.js raycaster caster picking. And then overload it in more specific cases like the tile layer. In this way, there is no need to add the picking in each provider of geometry.
@autra  It's clearer?

Comment (autra) [reply src/Core/Scheduler/Providers/TileProvider.js]: Somehow... 

> And then overload it in more specific cases like the tile layer

Where would you overload it then? I found that providers was a good place for that (because they constructs the objects, so they know how they should be picked). That being said I'm fine with the current version.

File: src/Core/Picking.js
Comment (gchoqueux) [inline src/Core/Picking.js]: The picking function shouldn't have as parameter the layer but the objects to pick.
Because this function must be usable without the notion of layer

File: src/Core/View.js
Comment (gchoqueux) [inline src/Core/View.js]: why `pickObjectsAt` isn't found in picking.js

File: src/Core/Scheduler/Providers/PointCloudProvider.js
Comment (gchoqueux) [inline src/Core/Scheduler/Providers/PointCloudProvider.js]: `pickObjectsAt` must be moved in geometry layer

**Review by gchoqueux — CHANGES_REQUESTED** [review]

File: src/Core/View.js
Comment (peppsac) [reply src/Core/View.js]: Because Picking.js is the implementation of the various picking methods (I'm also moving `raycastObjectsAt` in Picking.js)

`View.prototype.pickObjectsAt` is the boilerplate needed to use the pickings methods on a view.

File: src/Core/Scheduler/Providers/PointCloudProvider.js
Comment (peppsac) [reply src/Core/Scheduler/Providers/PointCloudProvider.js]: (small aside: please avoid this wording. "You must do this" is kinda rude)

There's no perfect place for this kind of code actually. PR #546 aimed to fix this but it couldn't be merged yet...

Here you're nitpicking on one line and your argument "Provider is to provide things" can be used to not move the picking code in `Layer.js` : "Layer should only contain generic things but Picking.pickPointsAt is a custom pick code".

So I suggest we move and work on improving #546 (or submit another PR that would solve this problem).

File: src/Core/View.js
Comment (peppsac) [reply src/Core/View.js]: See previous answer.

It's boilerplate code to ease picking usage from the View.

**Comment by peppsac** [comment]
BRanch updated.

**Comment by gchoqueux** [comment]
There is also `FeatureUtils` there is also foo as a method of picking

**Comment by peppsac** [comment]
> There is also FeatureUtils there is also foo as a method of picking

`FeatureUtils.filterFeaturesUnderCoordinate` is picking from a `Coordinates` object while this PR is focusing on objects selection from mouse coordinates.

Later, it could be interesting to extend the picking API to any kind of coordinates (pixels  or geographic) though.

**Comment by gchoqueux** [comment]
Without picking feature, so idea to expose a single entry point for querying isn't completed?

**Comment by peppsac** [comment]
> Without picking feature, so idea to expose a single entry point for querying isn't completed?

You can pick features at given window coordinates: `view.pickObjectsAt(mouse, wfsLayer)` will return all features under your mouse (or you can use Picking.pickObjectsAt(mouse, camera, someObject3d)` if you prefer, but for wfs the first form is easier since wfs objects are attached to TileMesh).

File: src/Core/View.js
Comment (gchoqueux) [reply src/Core/View.js]: ok

**Comment by gchoqueux** [comment]
and for 2D vector features?

**Comment by peppsac** [comment]
> and for 2D vector features?

This function can pick anything that is rendered as a Mesh. Does that answer your question?

File: src/Core/Scheduler/Providers/PointCloudProvider.js
Comment (gchoqueux) [reply src/Core/Scheduler/Providers/PointCloudProvider.js]: It can have several providers return the same type of object, THREE.Object3D for example, it makes more sense to move it in the construction of the geometric layer than to put picking code in  each provider that returns an object3D

**Comment by gchoqueux** [comment]
yes, the PR title and commit should be changed to : `unify picking mesh methods`

File: src/Core/Scheduler/Providers/PointCloudProvider.js
Comment (peppsac) [reply src/Core/Scheduler/Providers/PointCloudProvider.js]: > put  picking code in each provider that returns an object3D

- I'm doing this solely for the PointCloudProvider
- this line will go away as soon as #546 is merged
- I disagree that putting this pointcloud specific line is better in GeometryLayer constructor.

Will you block this PR for this?

Comment (gchoqueux) [reply src/Core/Scheduler/Providers/PointCloudProvider.js]: Are you ok with this [comment](https://github.com/iTowns/itowns/pull/635#discussion_r166295422)?

Otherwise, I agree, it makes more sense, first of all, to work and merge #546

Comment (peppsac) [reply src/Core/Scheduler/Providers/PointCloudProvider.js]: No.
I'd prefer to first merge this one, and then merge the more complicated one.

Comment (peppsac) [reply src/Core/Scheduler/Providers/PointCloudProvider.js]: Reminder: this PR doesn't remove nor add feature. It simply groups the picking-from-window-coordinate in the same module. I don't even see why we're arguing so much about it :(

**Comment by peppsac** [comment]
Added a small change : `Picking` module also returns the `layer.id` of the object (if it exists).

**Comment by peppsac** [comment]
So can I rebase this PR (and change the title) or are there any more comments?

**Comment by peppsac** [comment]
Rebased. And added a small fix: make sure we only display objects from the layer when doing the picking (see https://github.com/iTowns/itowns/pull/635/files#diff-11ecfe7e737c10e75920ac54de872437R8)

**Comment by peppsac** [comment]
Rebased since #637 was merged.

**Comment by peppsac** [comment]
@gchoqueux can you update your review of this PR please? (or dismiss it if you don't have time to update your review)

**Comment by gchoqueux** [comment]
there's [comment](https://github.com/iTowns/itowns/pull/635#discussion_r166301777) without reply 
and I add this previous [comment](https://github.com/iTowns/itowns/pull/637#issuecomment-364384095)

**Comment by peppsac** [comment]
> there's comment without reply

I've answered to all comments in this thread 6 days ago so...
If I missed something please be more specific.

>  and I add this previous comment

which I also answered.

**Comment by gchoqueux** [comment]
> I disagree that putting this pointcloud specific line is better in GeometryLayer constructor.

Why?

**Comment by peppsac** [comment]
> Why

Already answered as well in the comment  [above](https://github.com/iTowns/itowns/pull/635#discussion_r165957424): `Layer.js` should only contains generic code that applies to all layers, and `layer.pickObjectsAt = (view, mouse) => Picking.pickPointsAt(....)` is pointcloud generic.
And a more important argument: IMHO this dicussion is ±pointless because PR #546 should reorganize the code properly.
And see also the issue #182 that details the need for refactoring regarding Providers.

So unless you feel this is really important and you intend to block this PR because of this, I consider this a non issue.

File: examples/3dtiles.html
Comment (gchoqueux) [inline examples/3dtiles.html:158]: i'll be better than that goes on like this:
```js
layer.on('mouseMove', (event) => { 
    const intersects = event.intersects;
}
```

**Review by gchoqueux — CHANGES_REQUESTED** [review]

File: examples/3dtiles.html
Comment (peppsac) [reply examples/3dtiles.html:158]: Feel free to implement `.on()` for layers if you think it's useful, but that's not the scope of this PR.

**Comment by gchoqueux** [comment]
With approach `viewer.on`

* remove methods, `eventToViewCoords,  eventToNormalizedCoords, viewToNormalizedCoords and normalizedToViewCoords` in itowns's core and all users cases. Because their responsibilities are unloaded to the `viewerEvent`
* the same event `viewerEvent` is created and is dispatched to all listeners (which avoids redundant calculations)
* simplifies the code

```js
function picking(event) {
        var intersects = view.pickObjectsAt(view.eventToViewCoords(event), layer);
}
window.addEventListener('mousemove', picking, false);
```
become
```js
 function picking(event) {
        var intersects = event.intersects;
 }
 view.on('mousemove', picking, layer);
```
-------------------------------

**An example of partial implementation**
```js
const viewerListeners = {};
const raycaster = new Raycaster();
function viewerEvent(type, event, viewer) {
    const _event = {
        type,
        viewCoords: _eventCoords.set(event.offsetX, event.offsetY),
        event,
    };
    let normalizedCoords;
    let intersects;
    Object.defineProperty(
        _event,
        'normalizedCoords',
        {
            get: () => {
                if (!normalizedCoords) {
                    _eventCoords.x = 2 * (_event.viewCoords.x / viewer.camera.width) - 1;
                    _eventCoords.y = -2 * (_event.viewCoords.y / viewer.camera.height) + 1;
                    normalizedCoords = _eventCoords;
                }
                return normalizedCoords;
            },
        });
    Object.defineProperty(
    _event,
    'intersects',
        {
            get: () => {
                if (!intersects) {
                    if (_event.any instanceof GeometryLayer) {
                        raycaster.setFromCamera(_event.normalizedCoords, viewer.camera.camera3D);
                        intersects = raycaster.intersectObjects(_event.any.object3d.children, true);
                    } else {
                        intersects = [];
                    }
                }
                return intersects;
            },
        });

    return _event;
}

View.prototype.on = function _on(type, listener, any) {
    if (!viewerListeners[type]) {
        const domElement = this.mainLoop.gfxEngine.renderer.domElement;
        viewerListeners[type] = (event) => {
            this.dispatchEvent(viewerEvent(type, event, this));
        };
        if (type == 'click') {
            domElement.addEventListener('mousedown', viewerListeners[type]);
        } else if (type == 'mousemove') {
            domElement.addEventListener('mousemove', viewerListeners[type]);
        }
    }
    this.addEventListener(type, (event) => {
        event.any = any || { object3d: this.scene };
        listener(event);
    });
};
```

**Comment by peppsac** [comment]
Your proposed implementation won't work properly for pointclouds (raycasting in millions of points doesn't work well).
You want to replace the line:
```js
   intersects = raycaster.intersectObjects(_event.any.object3d.children, true);
```
With:
```js
   intersects  = _event.any.pickOjectsAt(viewer, viewer.eventToViewCoords(_event));
```
To get something that works with all types of GeometryLayers (and of course `pickOjectsAt` is exactly the function added by this PR).

And again, sorry but your `view.on(...)` proposal isn't relevant for this PR (already said so [earlier](https://github.com/iTowns/itowns/pull/635#discussion_r167595186)).

**Comment by gchoqueux** [comment]
The `view.on (...)` approach doesn't prevent appropriate processing for each type of `GeometryLayer`
Just define it in the `intersects` getter property

**Comment by autra** [comment]
Another PR for `view.on` please. We need to measure all the impacts with a working implementation (I'm not against it in itself btw). It's largely orthogonal with this changeset.

**Comment by peppsac** [comment]
@gchoqueux you missed the most important part of my comment : "And again, sorry but your view.on(...) proposal isn't relevant for this PR (already said so earlier).".
Maybe the solution here is not the best but it's an improvment over the current code.

**Comment by gchoqueux** [comment]
I didn't miss it. I will study this PR, so that it can be easily improved towards the proposal `view.on()`.

**Comment by autra** [comment]
@gchoqueux so you're blocking a PR for a proposal that is:
- unrelated to it
- won't even touch the same line numbers

why?

**Comment by peppsac** [comment]
PR rebased and commits squashed.
@gchoqueux there's nothing here that will prevent implementing the `.on(...)` mechanism.

File: src/Core/Picking.js
Comment (gchoqueux) [inline src/Core/Picking.js]: declare `depthRGBA` outside the function

File: src/Core/Picking.js
Comment (gchoqueux) [inline src/Core/Picking.js]: In cloud globe example, `oldParent` could be null on first picking

File: src/Core/View.js
Comment (gchoqueux) [inline src/Core/View.js]: could be replace by `for (const s of source)`?

File: src/Core/View.js
Comment (gchoqueux) [inline src/Core/View.js]: you convert all the layer'id string to layer?

File: src/Core/View.js
Comment (gchoqueux) [inline src/Core/View.js]: by example, if `source[i]`is point cloud layer the `raycast` method isn't appropriate

File: src/Core/View.js
Comment (gchoqueux) [inline src/Core/View.js]: ~~we could add layers filter parameter in the parent's method `pickObjectsAt`~~

File: src/Core/Picking.js
Comment (gchoqueux) [inline src/Core/Picking.js]: `View.prototype.viewToNormalizedCoords `could be a static method and used here (and avoid redundant code)

File: src/Core/Picking.js
Comment (gchoqueux) [inline src/Core/Picking.js:48]: Could you used `Object3D.traverseAncestors` ?

File: src/Core/Picking.js
Comment (gchoqueux) [inline src/Core/Picking.js:60]: it should also move `getPickingPositionFromDepth` in picking.js

File: src/Core/Picking.js
Comment (gchoqueux) [inline src/Core/Picking.js:96]: may be change function parameters to `renderViewTobuffer(scene, camera, x, y, width, height)` to clarify the scene switch in the viewer (and avoid `fakeScene`)

File: src/Core/Prefab/GlobeView.js
Comment (gchoqueux) [inline src/Core/Prefab/GlobeView.js:368]: it is necessary to browse the tree so we have the node in `picked [0] .object`?

File: src/Core/TileMesh.js
Comment (gchoqueux) [inline src/Core/TileMesh.js:106]: i don't know whether to check if `n` is an instance of `TileMesh`?

**Review by gchoqueux — CHANGES_REQUESTED** [review]

File: src/Core/Picking.js
Comment (peppsac) [reply src/Core/Picking.js]: This shouldn't be possible? Can you be more specific?

File: src/Core/Picking.js
Comment (peppsac) [reply src/Core/Picking.js]: Done.

File: src/Core/View.js
Comment (peppsac) [reply src/Core/View.js]: Yes.

File: src/Core/View.js
Comment (peppsac) [reply src/Core/View.js]: That's why we don't do raycast. We use instead the specialized `pickObjectsAt` method (see 10 lines above).

File: src/Core/View.js
Comment (peppsac) [reply src/Core/View.js]: Sorry I don't understand your comment.

File: src/Core/Picking.js
Comment (peppsac) [reply src/Core/Picking.js]: Instead I'll change the signature, to match the layer pickObjectsAt function:
```js
pickObjectsAt(view, mouse, object, target = []) { ... }
```

File: src/Core/Picking.js
Comment (peppsac) [reply src/Core/Picking.js:48]: No, because `traverseAncestors` can't be stopped as soon as you find what you were looking for.

File: src/Core/Picking.js
Comment (peppsac) [reply src/Core/Picking.js:60]: Well no.
For now these functions are only valid for tile based layer - so they'll stay in their respective View implementation until we provide a way to query this for ~all geometry layers.

File: src/Core/Picking.js
Comment (peppsac) [reply src/Core/Picking.js:96]: Let's discuss the change to this function in the dedicated PR: https://github.com/iTowns/itowns/pull/672

In the meantime this PR can use the fakeScene trick because it's harmless.

File: src/Core/Prefab/GlobeView.js
Comment (peppsac) [reply src/Core/Prefab/GlobeView.js:368]: I'm not sure to understand what you mean.
But we have to browse the tree to select/deselect tiles based on the picking result.

File: src/Core/TileMesh.js
Comment (peppsac) [reply src/Core/TileMesh.js:106]: It's not needed since `applyChangeState` checks for `changeState` existence.

File: src/Core/Picking.js
Comment (peppsac) [reply src/Core/Picking.js:96]: I simplified this code and removed the need for the `fakeScene`

File: src/Core/Picking.js
Comment (gchoqueux) [reply src/Core/Picking.js]: in pointcloud_globe.html example, in the first double click, `oldParent` is null

File: src/Core/View.js
Comment (gchoqueux) [reply src/Core/View.js]: ok

File: src/Core/View.js
Comment (gchoqueux) [reply src/Core/View.js]: forget the previous comment

Why geometry layer doesn't have default `pickObjectsAt` method?

Comment (peppsac) [reply src/Core/View.js]: All GeometryLayer have a default picking method (see https://github.com/iTowns/itowns/pull/635/files#diff-61e012c3e00bdc510be33d9eac684b2aR80)

But the method in View support picking for all objects (for instance raw THREE.Object3D) so we need to use `Picking.pickObjectsAt` as a fallback.

File: src/Core/View.js
Comment (gchoqueux) [inline src/Core/View.js]: can be a single loop is sufficient?

File: src/Core/Picking.js
Comment (gchoqueux) [reply src/Core/Picking.js:60]: `pickTilesAt` is only valid for tile based layer?

File: src/Core/Prefab/GlobeView.js
Comment (gchoqueux) [reply src/Core/Prefab/GlobeView.js:368]: in other PR: it will later be necessary to keep the selection to avoid traverse the tree to update the nodes

File: src/Core/TileMesh.js
Comment (gchoqueux) [reply src/Core/TileMesh.js:106]: yes

File: src/Core/Picking.js
Comment (peppsac) [reply src/Core/Picking.js:60]: yep

File: src/Core/View.js
Comment (peppsac) [reply src/Core/View.js]: It could be done in a single loop but I think it helpes readability to separate the 2 loops.

**Comment by peppsac** [comment]
I think I've answered/fixed to all your comments. Did I miss something?

File: src/Core/Picking.js
Comment (gchoqueux) [reply src/Core/Picking.js:60]: your proposed picking allows you to get position for which type of geometry layers?

File: src/Core/View.js
Comment (gchoqueux) [reply src/Core/View.js]: With a comment will be more readable
and could you replace by `for (const source of sources)`

File: src/Core/View.js
Comment (gchoqueux) [inline src/Core/View.js]: use of a target isn't necessary

```js
const obj = Picking.pickObjectsAt(
                    this,
                    mouse,
                    parentLayer.object3d);
```

File: src/Core/View.js
Comment (gchoqueux) [reply src/Core/View.js]: > for instance raw THREE.Object3D 

Should be explain in the title commit because it doesn't unify the picking position/coordinates and features (in raster)

File: src/Core/View.js
Comment (gchoqueux) [inline src/Core/View.js]: returns only object3D?

File: src/Core/Scheduler/Providers/PointCloudProvider.js
Comment (gchoqueux) [reply src/Core/Scheduler/Providers/PointCloudProvider.js]: add a temporary comment : to indicate that it'll not stay in the provider

File: src/Core/View.js
Comment (peppsac) [reply src/Core/View.js]: Fixed both comments.

Comment (peppsac) [reply src/Core/View.js]: And also adapted pickObjectsAt to accept either mouse coords or a (Mouse|Touch)Event

File: src/Core/View.js
Comment (peppsac) [reply src/Core/View.js]: See the @return documentation for the returned values.

**Comment by peppsac** [comment]
I'll change the commit title when squashing the commits.

File: src/Core/View.js
Comment (gchoqueux) [inline src/Core/View.js]: if a geometric layer is attached to another geometric layer.
We could affect it the reference of `parentLayer.object3d` in the `GeometryLayer.prototype.attach` function
`layer.object3d = this.object3d`

Who would avoid the party with `it hasn't: this layer is attached to another one`

WDYT?

Comment (peppsac) [reply src/Core/View.js]: In other places of itowns, layer.object3d is the object used to add layer's mesh to the scene.

So I don't think it make sense for now to store parent layer's object3D here.

**Comment by gchoqueux** [comment]
OK rebase, squash and re word commit then we merge

**Comment by peppsac** [comment]
Done

**Review by gchoqueux — APPROVED** [review]
thanks

**Pull request merged**
